### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by ancr7**

### CODE REVIEW

File: src/main/java/Assignment2/Repo.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.ArrayList;
+
+public class Repo {
+
+  FileOutputStream fos;
```
Comment (hardikbohra) [inline src/main/java/Assignment2/Repo.java]: always use access identifiers wherever required (public/private/protected)

File: src/main/java/Assignment2/Repo.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  FileInputStream fis;
+
+  Repo() {
+  }
+
+  void write(ArrayList<StudentModel> studentList) throws Exception {
+    fos = new FileOutputStream(new File("DB.txt"));
```
Comment (hardikbohra) [inline src/main/java/Assignment2/Repo.java]: try using singleton class concept here.

File: src/main/java/Assignment2/StudentService.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+      System.out.println("Invalid address");
+      return false;
+    }
+    // check age empty
+    if (age.isEmpty()) {
+      System.out.println("Invalid age");
+      return false;
+    }
+    // check courses empty
+    if (courses.isEmpty()) {
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java]: try reducing number of conditions. use && and || operators

File: src/main/java/Assignment2/StudentService.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      System.out.println("Invalid courses");
+      return false;
+    }
+    // validating name
+    if (!name.matches("^[ A-Za-z]+$")) {
+      System.out.println("Invalid name");
+      return false;
+    }
+    // validating rollNo
+    if (!rollNo.matches("[0-9]*")) {
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java]: change error message

Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java]: what if someone is passing the age as 0?

File: src/main/java/Assignment2/StudentService.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    // validating age
+    if (!age.matches("[0-9]*")) {
+      System.out.println("Invalid quantity");
+      return false;
+    }
+    // validating course
+    String[] course = courses.split(" ");
+    HashSet<String> uniqueCourses = new HashSet<String>();
+    for (String i : course) {
+      if (i.length() > 1 || !i.matches("^[ A-F]+$")) {
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java]: use enum for courses.

File: src/main/java/Assignment2/StudentService.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    model.courses = courses.replaceAll(" ", "").toCharArray();
+    // addStudent(model)
+    addStudent(model);
+  }
+
+  void sortStudentList() {
+    // check if state changed
+    if (!isStateChanged) return;
+    // if yes then sort
+    if (currentType == sortOption.NAME) {
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java]: use equals for string match

File: src/main/java/Assignment2/StudentService.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+      if (isAscending) studentList
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java]: even for single statement in if block, use {} always for better readable code

File: src/main/java/Assignment2/StudentService.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+          .sort(Comparator.comparing(StudentModel::getName).thenComparing(StudentModel::getRollNo));
+      else studentList.sort(
+          Comparator.comparing(StudentModel::getName).thenComparing(StudentModel::getRollNo)
+              .reversed());
+    }
+    if (currentType == sortOption.AGE) {
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java]: think about switch case implementation

File: src/main/java/Assignment2/StudentService.java [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+    if (currentType == sortOption.ROLL_NO) {
+      if (isAscending) studentList.sort(Comparator.comparing(StudentModel::getRollNo));
+      else studentList.sort(Comparator.comparing(StudentModel::getRollNo).reversed());
+    }
+    isStateChanged = false;
+  }
+
+  void sortPreference(sortOption sortType, boolean isAscending) {
+    // same then nothing
+    if (sortType == currentType && isAscending == this.isAscending) return;
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java:161]: here also, == should not be used for string datatype comparisons

File: src/main/java/Assignment2/StudentService.java [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    // print data in studentList
+    for (StudentModel student : studentList) {
+      student.print();
+    }
+  }
+
+  void deleteStudent(String rollNo) {
+    // validating input
+    if (rollNo.isEmpty() || !rollNo.matches("[0-9]*")) {
+      System.out.println("Invalid quantity");
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java:179]: change error message

File: src/main/java/Assignment2/StudentController.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+      // validating input
+      if (ChoiceString.isEmpty() || !ChoiceString.matches("[0-9]*")
+          || Integer.parseInt(ChoiceString) > 5) {
+        System.out.println("Invalid choice");
+        continue;
+      }
+      int choice = Integer.parseInt(ChoiceString);
+      switch (choice) {
+        case 1: { // add data
+          String name, address, courses, age, rollno;
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentController.java]: move each and every case body into private methods. one method should not be this much long.

File: src/main/java/Assignment2/StudentService.java [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+      if (studentList.get(i).rollNo == Integer.parseInt(rollNo)) {
+        studentList.remove(i);
+        rollNoInRepo.remove(Integer.parseInt(rollNo));
+        isStateChanged = true;
+        break;
+      }
+    }
+  }
+
+  void saveState() {
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java:198]: every call of saveState() and getState() is creating repo object means db object. prevent this duplicate object creation.

File: src/test/java/Assignment2/StudentServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class StudentServiceTest {
+
+  @Test
+  void checkIfNameEmpty() {
+    StudentService studentService = new StudentService();
```
Comment (hardikbohra) [inline src/test/java/Assignment2/StudentServiceTest.java]: don't create studentService object everytime.
use @Before Annotation

File: src/test/java/Assignment2/StudentServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    assertEquals(false, studentService.validInput("abc", "12", "abs scs", "12", "A C D"));
+  }
+
+  @Test
+  void checkIfRollNoAlreadyExistWhenAddingStudent() {
+    StudentService studentService = new StudentService();
+    studentService.processInput("abc", "12", "x y z", "12", "A B C D");
+    studentService.processInput("def", "12", "m n o", "22", "C D E F");
+    assertEquals(1, studentService.studentList.size());
```
Comment (hardikbohra) [inline src/test/java/Assignment2/StudentServiceTest.java:88]: shouldn't the size value be 2?

File: src/test/java/Assignment2/StudentServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+  @Test
+  @DisplayName("When rollNo format is wrong")
+  void checkInputRollNo() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "2F", "x y z", "21", "A B C D"));
+  }
+
+  @Test
+  @DisplayName("When Age format is wrong")
+  void checkInputAge() {
```
Comment (hardikbohra) [inline src/test/java/Assignment2/StudentServiceTest.java:64]: check when age and rollnumber passed as 00

**Review by hardikbohra — CHANGES_REQUESTED** [review]

File: src/test/java/Assignment2/StudentServiceTest.java (hunk 14 above)
Comment (ancr7) [reply src/test/java/Assignment2/StudentServiceTest.java:88]: Since both has same rollNo only first entry will be stored second will be discarded.

Comment (hardikbohra) [reply src/test/java/Assignment2/StudentServiceTest.java:88]: ohhk correct

File: src/main/java/Assignment2/Repo.java (hunk 1 above)
Comment (ancr7) [reply src/main/java/Assignment2/Repo.java]: got it, done.

File: src/main/java/Assignment2/Repo.java (hunk 2 above)
Comment (ancr7) [reply src/main/java/Assignment2/Repo.java]: done

File: src/main/java/Assignment2/StudentController.java (hunk 11 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentController.java]: done.

File: src/main/java/Assignment2/StudentService.java (hunk 3 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java]: done.

File: src/main/java/Assignment2/StudentService.java (hunk 4 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java]: done.

Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java]: added the check,done.

File: src/main/java/Assignment2/StudentService.java (hunk 6 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java]: I use == for comparing enum.

File: src/main/java/Assignment2/StudentService.java (hunk 7 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java]: done.

File: src/main/java/Assignment2/StudentService.java (hunk 8 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java]: done.

File: src/main/java/Assignment2/StudentService.java (hunk 9 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java:161]: I use == for comparing enum.

File: src/main/java/Assignment2/StudentService.java (hunk 10 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java:179]: done.

File: src/main/java/Assignment2/StudentService.java (hunk 12 above)
Comment (ancr7) [reply src/main/java/Assignment2/StudentService.java:198]: done.

File: src/test/java/Assignment2/StudentServiceTest.java (hunk 15 above)
Comment (ancr7) [reply src/test/java/Assignment2/StudentServiceTest.java:64]: added the check, done.

File: src/test/java/Assignment2/StudentServiceTest.java (hunk 13 above)
Comment (ancr7) [reply src/test/java/Assignment2/StudentServiceTest.java]: done.

File: src/main/java/Assignment2/StudentService.java [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+      }
+      return false;
+    }
+  }
+
+  StudentService() {
+    studentList = new ArrayList<>();
+    isStateChanged = false;
+    isAscending = true;
+    currentType = NAME;
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java:41]: use enum value to initialize currentType

File: src/main/java/Assignment2/StudentService.java [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+      return false;
+    }
+    // validating name
+    if (name.isEmpty() || !name.matches("^[ A-Za-z]+$")) {
+      System.out.println("Invalid name format");
+      return false;
+    }
+    // validating rollNo
+    if (rollNo.isEmpty() || !rollNo.matches("[0-9]*")) {
+      System.out.println("Invalid quantity format");
```
Comment (hardikbohra) [inline src/main/java/Assignment2/StudentService.java]: change error message

**Review by hardikbohra — APPROVED** [review]

**Pull request merged**